src/Workspaces/Core/Portable/Classification/AbstractClassificationService.cs

```diff
@@ -64,6 +72,45 @@ public async Task AddSemanticClassificationsAsync(Document document, TextSpan te
             }
         }
 
+        private static bool IsFullyLoaded(Document document, CancellationToken cancellationToken)
+        {
+            var workspaceStatusService = document.Project.Solution.Workspace.Services.GetRequiredService<IWorkspaceStatusService>();
+
+            // Importantly, we do not await/wait on the fullyLoadedStateTask.  We do not want to ever be waiting on work
+            // that may end up touching the UI thread (As we can deadlock if GetTagsSynchronous waits on us).  Instead,
+            // we only check if the Task is completed.  Prior to that we will assume we are still loading.  Once this
+            // task is completed, we know that the WaitUntilFullyLoadedAsync call will have actually finished and we're
+            // fully loaded.
+            var isFullyLoadedTask = workspaceStatusService.IsFullyLoadedAsync(cancellationToken);
+            var isFullyLoaded = isFullyLoadedTask.IsCompleted && isFullyLoadedTask.GetAwaiter().GetResult();
+            return isFullyLoaded;
+        }
+
+        private static async Task<bool> TryGetCachedClassificationsAsync(
+            Document document, TextSpan textSpan, ArrayBuilder<ClassifiedSpan> result,
+            RemoteHostClient client, bool isFullyLoaded, CancellationToken cancellationToken)
```

Comment:
importantly, this work should not be pushed into the OOP client.  the idea here is that we want to make this request *without* syncing over the full solution.  so we must make the decision on the client to checked for cached results first, and if not go to normal results.